### PR TITLE
Support custom shapes for dummy inputs

### PR DESCRIPTION
**Pull request by fxmarty**
As per title, this is part of the effort to make ONNX export more reliable (reference https://github.com/huggingface/optimum/issues/503 )

Next PRs will come to:
* Allow custom shapes during `torch.onnx.export`
* More extensive testing of exported models with `optimum.exporters`
* Include log information in the ONNX export, some info on the supported/unsupported shapes / inputs

## Before submitting
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/exporters/onnx/config.py [hunk 1]
```diff
@@ -85,7 +89,7 @@ def inputs(self) -> Mapping[str, Mapping[int, str]]:
 
         return common_inputs
 
-    def create_dummy_input_generator_classes(self):
+    def _create_dummy_input_generator_classes(self, **kwargs) -> List["DummyInputGenerator"]:
```
Comment (michaelbenayoun) [inline optimum/exporters/onnx/config.py:92]: None of the dummy input generators need `**kwargs` in this case?

File: optimum/exporters/onnx/model_configs.py [hunk 2]
```diff
@@ -304,7 +305,7 @@ class BartOnnxConfig(TextSeq2SeqOnnxConfig):
         },
     )
 
-    def create_dummy_input_generator_classes(self):
+    def _create_dummy_input_generator_classes(self, **kwargs) -> List["DummyInputGenerator"]:
```
Comment (michaelbenayoun) [inline optimum/exporters/onnx/model_configs.py:308]: Same question

File: tests/utils/test_dummpy_input_generators.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                "num_choices": DUMMY_SHAPES["num_choices"],
+                "sequence_length": DUMMY_SHAPES["sequence_length"],
+            }
+        )
+    )
+    def test_text_models(self, model_name: str, batch_size: int, num_choices: int, sequence_length: int):
+        # isn't this very verbose?
+        config = AutoConfig.from_pretrained(model_name)
+        normalized_config_class = NormalizedConfigManager.get_normalized_config_class(config.model_type)
+        normalized_config = normalized_config_class(config)
```
Comment (michaelbenayoun) [inline tests/utils/test_dummpy_input_generators.py:79]: Yes you are right it is!
We can create a `from_pretrained` method to `NormalizedConfig` handling all of this!

File: optimum/exporters/onnx/base.py [hunk 4]
```diff
@@ -249,24 +249,44 @@ def ordered_inputs(self, model: Union["PreTrainedModel", "TFPreTrainedModel"]) -
...
+        dummy_inputs_generators = self._create_dummy_input_generator_classes(**kwargs)
+
         dummy_inputs = {}
         for input_name in self.inputs:
             input_was_inserted = False
-            for dummy_input_gen in self.dummy_inputs_generators:
+            for dummy_input_gen in dummy_inputs_generators:
                 if dummy_input_gen.supports_input(input_name):
-                    dummy_inputs[input_name] = dummy_input_gen.generate(input_name, framework=framework)
+                    dummy_inputs[input_name] = dummy_input_gen.generate(input_name, framework=framework, **kwargs)
```
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:289]: Why does `generate` need to have `**kwargs`?

File: optimum/exporters/onnx/config.py (hunk 1 above)
Comment (fxmarty) [reply optimum/exporters/onnx/config.py:95]: They do. Should be fixed in https://github.com/huggingface/optimum/pull/522/commits/4ec889317b1bb3528191632545c997b0bc0e8412 , can you check again if it looks good?

File: optimum/exporters/onnx/model_configs.py (hunk 2 above)
Comment (fxmarty) [reply optimum/exporters/onnx/model_configs.py:311]: cf above

File: tests/utils/test_dummpy_input_generators.py (hunk 3 above)
Comment (fxmarty) [reply tests/utils/test_dummpy_input_generators.py:79]: Ok let's do it in an other PR

File: optimum/exporters/onnx/base.py (hunk 4 above)
Comment (fxmarty) [reply optimum/exporters/onnx/base.py]: It does not. Thanks for noticing!

File: optimum/exporters/onnx/base.py [hunk 5]
```diff
@@ -249,24 +249,44 @@ def ordered_inputs(self, model: Union["PreTrainedModel", "TFPreTrainedModel"]) -
...
+                The height to use in the dummy inputs for vision tasks.
+            num_channels (`int`, *optional*, defaults to 3):
+                The number of channels to use in the dummpy inputs for vision tasks.
+            feature_size (`int`, *optional*, defaults to 80):
+                The number of features to use in the dummpy inputs for audio tasks in case it is not raw audio.
+                This is for example the number of STFT bins or MEL bins.
+            nb_max_frames (`int`, *optional*, defaults to 3000):
+                The number of frames to use in the dummpy inputs for audio tasks in case the input is not raw audio.
+            audio_sequence_length (`int`, *optional*, defaults to 16000):
+                The number of frames to use in the dummpy inputs for audio tasks in case the input is raw audio.
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/exporters/onnx/base.py]: Ok so apparently docbuilder does not support f-strings. I used the trick from `modeling_XXX.py` from transformers where they dynamically generate the examples. Let me know what you think: https://github.com/huggingface/optimum/pull/522/commits/0c6959e9e6ea99053a5e04f33baa025c1cfc8d7e

**Comment by fxmarty** [comment]
I don't quite understand why those are needed

https://github.com/huggingface/optimum/blob/3d0a39c9c955d18989209fa24d30187850cab856/optimum/exporters/onnx/base.py#L362

https://github.com/huggingface/optimum/blob/3d0a39c9c955d18989209fa24d30187850cab856/optimum/exporters/onnx/model_configs.py#L497

https://github.com/huggingface/optimum/blob/3d0a39c9c955d18989209fa24d30187850cab856/optimum/exporters/onnx/model_configs.py#L747

and whether they should be in the doc / have docstrings

**Comment by fxmarty** [comment]
Ok, so I understand we don't need to include in the doc the reimplemented children `generate_dummy_inputs`.

File: optimum/utils/normalized_config.py [hunk 6]
```diff
@@ -94,3 +94,123 @@ def __getattr__(self, attr_name):
...
+T5LikeNormalizedTextConfig = NormalizedTextConfig.with_args(
+    num_attention_heads="num_heads",
+    hidden_size="d_model",
+)
+WhisperLikeNormalizedTextConfig = NormalizedTextConfig.with_args(
+    hidden_size="d_model",
+)
+
+
+class NormalizedConfigManager:
```
Comment (michaelbenayoun) [inline optimum/utils/normalized_config.py:113]:
```suggestion
class NormalizedConfigsManager:
```

**Review by michaelbenayoun — APPROVED** [review]
LGTM

**Pull request merged**